qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -317,39 +317,38 @@ def _obtain_swaps(self, front_layer, current_layout):
 
         return candidate_swaps
 
+    def _compute_cost(self, layer, layout):
+        cost = 0
+        for node in layer:
+            cost += self.coupling_map.distance(layout[node.qargs[0]], layout[node.qargs[1]])
+        return cost
+
     def _score_heuristic(self, heuristic, front_layer, extended_set, layout, swap_qubits=None):
         """Return a heuristic score for a trial layout.
 
         Assuming a trial layout has resulted from a SWAP, we now assign a cost
         to it. The goodness of a layout is evaluated based on how viable it makes
         the remaining virtual gates that must be applied.
         """
+        first_cost = self._compute_cost(front_layer, layout)
         if heuristic == "basic":
-            if len(front_layer) > 1:
-                return self.coupling_map.distance_matrix[
-                    tuple(zip(*([layout[q] for q in node.qargs] for node in front_layer)))
-                ].sum()
-            elif len(front_layer) == 1:
-                return self.coupling_map.distance(*(layout[q] for q in list(front_layer)[0].qargs))
-            else:
-                return 0
```

Comment:
This was an optimization I did in #5403 to try and improve the speed of the cost computation for larger coupling maps. Using the numpy sum was slightly faster in those cases IIRC (it looks like I didn't put numbers in the PR so it might not have actually made a difference). I'm just leaving a comment here to make a note of this for the future if we hit scaling problems here again to try this again and benchmark it that time.

Reply:
I can look further into it, I didn't like how the code was calling itself and also was quite crypt. In my profiling experiments I did not see any measurable difference, but I suppose we could use it on the function that effectively computes the cost.

Reply:
Yeah, what's in this pr is way more legible. I think it's fine like this (which is why I approved and merged the pr) I'm not actually convinced the numpy sum() call made a noticeable difference (none of the asv benchmarks show a difference and I didn't post numbers in the pr) and it's quite dense to read. If the sum really starts to be a bottleneck we should probably just look at using cython or another compiled implementation instead.